### ISSUE TITLE
repairs repeatedly postponed and stuck

### ISSUE THREAD

**Issue by Gappa88** [thelastpickle/cassandra-reaper#863]
**Reaper v2.0.3
Cassandra v3.11.5.1**

every day I run a repair on a single keyspace and since few weeks ago repairs never end. 
The following is the information table taken from the reaper's dashboard:

```

ID | 00000000-0000-0177-0000-000000000000
-- | --
Owner | g
Cause | g
Last event | postponed repair segment 00000000-0000-c696-0000-000000000000 because one of the hosts (xx.xx.xx.xx) was already involved in a repair
Start time | March 9, 2020 10:45 AM
End time |  
Pause time |  
Duration | 22 hours 17 minutes 10 seconds
Segment count | 136
Segment repaired | 67
Intensity | 0.8999999761581421
Repair parallelism | PARALLEL
Incremental repair | false
Repair threads | 1
Nodes |  
Datacenters | DC1
Blacklist |  
Creation time | March 9, 2020 10:45 AM
Available metrics(can require a full run before appearing) | io.cassandrareaper.service.RepairRunner.repairProgress. mycluster.mkphistory.00000000000000070000000000000000io.cassandrareaper.service.RepairRunner.segmentsDone. mycluster.mkphistory.00000000000000070000000000000000io.cassandrareaper.service.RepairRunner.segmentsTotal. mycluster.mkphistory.00000000000000070000000000000000io.cassandrareaper.service.RepairRunner.millisSinceLastRepair. mycluster.mkphistory.00000000000000070000000000000000
```

I also noted the very same message in the reaper's log repeated infinite times:
```
INFO   [ mycluster:00000000-0000-0177-0000-000000000000:00000000-0000-c696-0000-000000000000] i.c.s.RepairRunner - postponed repair segment 00000000-0000-c696-0000-000000000000 because one of the hosts (xx.xx.xx.xx) was already involved in a repair
INFO   [ mycluster:00000000-0000-0177-0000-000000000000:00000000-0000-c696-0000-000000000000] i.c.s.RepairRunner - postponed repair segment 00000000-0000-c696-0000-000000000000 because one of the hosts (xx.xx.xx.xx) was already involved in a repair
INFO   [ mycluster:00000000-0000-0177-0000-000000000000:00000000-0000-c696-0000-000000000000] i.c.s.RepairRunner - postponed repair segment 00000000-0000-c696-0000-000000000000 because one of the hosts (xx.xx.xx.xx) was already involved in a repair
INFO   [ mycluster:00000000-0000-0177-0000-000000000000:00000000-0000-c696-0000-000000000000] i.c.s.RepairRunner - postponed repair segment 00000000-0000-c696-0000-000000000000 because one of the hosts (xx.xx.xx.xx) was already involved in a repair
INFO   [ mycluster:00000000-0000-0177-0000-000000000000:00000000-0000-c696-0000-000000000000] i.c.s.RepairRunner - postponed repair segment 00000000-0000-c696-0000-000000000000 because one of the hosts (xx.xx.xx.xx) was already involved in a repair
INFO   [ mycluster:00000000-0000-0177-0000-000000000000:00000000-0000-c696-0000-000000000000] i.c.s.RepairRunner - postponed repair segment 00000000-0000-c696-0000-000000000000 because one of the hosts (xx.xx.xx.xx) was already involved in a repair
INFO   [ mycluster:00000000-0000-0177-0000-000000000000:00000000-0000-c696-0000-000000000000] i.c.s.RepairRunner - postponed repair segment 00000000-0000-c696-0000-000000000000 because one of the hosts (xx.xx.xx.xx) was already involved in a repair
INFO   [ mycluster:00000000-0000-0177-0000-000000000000:00000000-0000-c696-0000-000000000000] i.c.s.RepairRunner - postponed repair segment 00000000-0000-c696-0000-000000000000 because one of the hosts (xx.xx.xx.xx) was already involved in a repair
INFO   [ mycluster:00000000-0000-0177-0000-000000000000:00000000-0000-c696-0000-000000000000] i.c.s.RepairRunner - postponed repair segment 00000000-0000-c696-0000-000000000000 because one of the hosts (xx.xx.xx.xx) was already involved in a repair
```

Few weeks ago this repair lasts just a couple of hours launched with 4 threads. I tried di decrease the number of thread employed in the repair but the result hasn't changed and the repair still stuck.

I also tried a rolling restart (I restarted also the reaper) without success.

Do you have any idea about this behavior?

**Comment by therb1** [comment]
Please show commands output:
```
nodetool status
nodetool describecluster
```
Mask youre personal data if command output like ip addresses.

**Comment by Gappa88** [comment]
thanks @therb1 for the answer.

here what you requested:
```
# nodetool status
Datacenter: DC1
===============
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address        Load       Tokens       Owns    Host ID                               Rack
UN  xxxxxxxxxxxxx  2.65 TiB   64           ?       0c59a9e1-2057-4f40-93b9-89a6ee855b07  r1
UN  xxxxxxxxxxxxx  2.94 TiB   64           ?       49135ac7-2b1c-42a6-b34c-02b7a78efdd4  r1
UN  xxxxxxxxxxxxx  2.88 TiB   64           ?       11497fc6-3f88-42fd-a4a2-551f4bfbc66c  r1
UN  xxxxxxxxxxxxx  2.73 TiB   64           ?       12612de5-0aa3-428b-9adc-1d9dbb3e2ae7  r1
UN  xxxxxxxxxxxxx  2.43 TiB   64           ?       cd284aec-e1ba-460c-b9a5-c5c3e7036fd5  r1
UN  xxxxxxxxxxxxx  2.78 TiB   64           ?       7d86fa74-485a-4874-ab8d-7ce6a0210077  r1
```

```
# nodetool describecluster
Cluster Information:
        Name: myCluster
        Snitch: org.apache.cassandra.locator.GossipingPropertyFileSnitch
        DynamicEndPointSnitch: enabled
        Partitioner: org.apache.cassandra.dht.Murmur3Partitioner
        Schema versions:
                3f39b4a2-a8cd-301e-8c1c-9732bc1e95d7: [xxxxxxxxxxxxx, xxxxxxxxxxxxx, xxxxxxxxxxxxx, xxxxxxxxxxxxx, xxxxxxxxxxxxx, xxxxxxxxxxxxx]
```

**Comment by therb1** [comment]
Have you tried to start the repair locally on the cassandra nodes?

**Comment by Gappa88** [comment]
@therb1 no, I did not try to start the repair locally on each nodes because I have always used cassandra-reaper to do that. Actually, I'm not sure to be able to do that, If you have any advice I will try to launch a repair locally

**Comment by powellchristoph** [comment]
I have also run into this. Segements postponed for 4 days and when I look on the node that the log line says its running a repair on I see the node has completed the repaired segment days ago.

**Comment by Gappa88** [comment]
@powellchristoph , do you have a solution for this?

**Comment by powellchristoph** [comment]
I try to abort the segment, both through the UI and with `spreader`. The command run successfully but the segment stays.

```
Mar 16 14:13:17 cassandra-reaper-n01 cassandra-reaper[12757]: INFO   [2020-03-16 14:13:17,831] [aux1:ac0fbb90-62e0-11ea-beea-29aaae9654a6:ac10cd1a-62e0-11ea-beea-29aaae9654a6] i.c.s.RepairRunner - postponed repair segment ac10cd1a-62e0-11ea-beea-29aaae9654a6 because segment ac131704-62e0-11ea-beea-29aaae9654a6 is running on host 34.232.27.8
Mar 16 14:13:20 cassandra-reaper-n01 cassandra-reaper[12757]: INFO   [2020-03-16 14:13:20,071] [aux1:ac0fbb90-62e0-11ea-beea-29aaae9654a6:ac16e7c4-62e0-11ea-beea-29aaae9654a6] i.c.s.RepairRunner - postponed repair segment ac16e7c4-62e0-11ea-beea-29aaae9654a6 because segment ac131704-62e0-11ea-beea-29aaae9654a6 is running on host 34.232.27.8
Mar 16 14:13:21 cassandra-reaper-n01 cassandra-reaper[12757]: INFO   [2020-03-16 14:13:21,414] [dw-281 - POST /repair_run/ac0fbb90-62e0-11ea-beea-29aaae9654a6/segments/abort/ac131704-62e0-11ea-beea-29aaae9654a6] i.c.s.CassandraStorage - Trying to release lead on segment ac131704-62e0-11ea-beea-29aaae9654a6 for instance a4d65cdf-e0c9-4683-a913-4de421604999
Mar 16 14:13:21 cassandra-reaper-n01 cassandra-reaper[12757]: INFO   [2020-03-16 14:13:21,414] [dw-281 - POST /repair_run/ac0fbb90-62e0-11ea-beea-29aaae9654a6/segments/abort/ac131704-62e0-11ea-beea-29aaae9654a6] i.c.s.CassandraStorage - Released lead on segment ac131704-62e0-11ea-beea-29aaae9654a6
Mar 16 14:13:25 cassandra-reaper-n01 cassandra-reaper[12757]: INFO   [2020-03-16 14:13:25,345] [aux1:ac0fbb90-62e0-11ea-beea-29aaae9654a6:ac17aaec-62e0-11ea-beea-29aaae9654a6] i.c.s.RepairRunner - postponed repair segment ac17aaec-62e0-11ea-beea-29aaae9654a6 because segment ac131704-62e0-11ea-beea-29aaae9654a6 is running on host 34.232.27.8
Mar 16 14:13:28 cassandra-reaper-n01 cassandra-reaper[12757]: INFO   [2020-03-16 14:13:28,063] [aux1:ac0fbb90-62e0-11ea-beea-29aaae9654a6:ac1adf82-62e0-11ea-beea-29aaae9654a6] i.c.s.RepairRunner - postponed repair segment ac1adf82-62e0-11ea-beea-29aaae9654a6 because segment ac131704-62e0-11ea-beea-29aaae9654a6 is running on host 34.232.27.8
Mar 16 14:13:30 cassandra-reaper-n01 cassandra-reaper[12757]: INFO   [2020-03-16 14:13:30,678] [aux1:ac0fbb90-62e0-11ea-beea-29aaae9654a6:ac1a6a3b-62e0-11ea-beea-29aaae9654a6] i.c.s.RepairRunner - postponed repair segment ac1a6a3b-62e0-11ea-beea-29aaae9654a6 because segment ac131704-62e0-11ea-beea-29aaae9654a6 is running on host 34.232.27.8
```

**Comment by powellchristoph** [comment]
Here is more information. A segment hangs and then appears to eventually block all remaining repairs involving that node.

Cassandra Reaper version: 2.0.3
Cassandra Version: 3.11.5

This is a multi-datacenter cluster setup with a Reaper in each datacenter and a shared Cassandra backend.

I removed all schedules, clusters and dropped the keyspace on the Cassandra-backend. I recreated the keyspace and let Reaper do a fresh migration of its tables. Started a fresh repair and had a section hang very quickly.

I attempt to abort the "stuck" segment `a1f35aec-67a0-11ea-93a2-b332b9300b0a`.
```
73001 INFO   [aux1:a1e5c640-67a0-11ea-93a2-b332b9300b0a:a1f35aec-67a0-11ea-93a2-b332b9300b0a] i.c.s.SegmentRunner - Repair command 3249 on segment a1f35aec-67a0-11ea-93a2-      b332b9300b0a returned with state STARTED
73002 INFO   [aux1:a1e5c640-67a0-11ea-93a2-b332b9300b0a:a1f35aec-67a0-11ea-93a2-b332b9300b0a] i.c.s.SegmentRunner - Repair command 3249 on segment a1f35aec-67a0-11ea-93a2-      b332b9300b0a has been cancelled while running
73003 INFO   [aux1:a1e5c640-67a0-11ea-93a2-b332b9300b0a:a1f35aec-67a0-11ea-93a2-b332b9300b0a] i.c.s.SegmentRunner - Postponing segment a1f35aec-67a0-11ea-93a2-b332b9300b0a
73004 INFO   [aux1:a1e5c640-67a0-11ea-93a2-b332b9300b0a:a1f4936c-67a0-11ea-93a2-b332b9300b0a] i.c.s.RepairRunner - postponed repair segment a1f4936c-67a0-11ea-93a2-b332b93      00b0a because segment a1eccb23-67a0-11ea-93a2-b332b9300b0a is running on host 34.232.27.8
73005 INFO   [aux1:a1e5c640-67a0-11ea-93a2-b332b9300b0a:a1f4936c-67a0-11ea-93a2-b332b9300b0a] i.c.s.SegmentRunner - Cannot run segment a1f4936c-67a0-11ea-93a2-b332b9300b0a       for repair a1e5c640-67a0-11ea-93a2-b332b9300b0a at the moment. Will try again later
73006 INFO   [aux1:a1e5c640-67a0-11ea-93a2-b332b9300b0a] i.c.s.RepairRunner - Running segment for range (-1597631391510650955,202824120446520166]
73007 INFO   [aux1:a1e5c640-67a0-11ea-93a2-b332b9300b0a] i.c.s.RepairRunner - Next segment to run : a1ee78df-67a0-11ea-93a2-b332b9300b0a
73008 INFO   [aux1:a1e5c640-67a0-11ea-93a2-b332b9300b0a:a1f35aec-67a0-11ea-93a2-b332b9300b0a] i.c.s.CassandraStorage - Trying to release lead on segment a1f35aec-67a0-11ea      -93a2-b332b9300b0a for instance c57b7ae7-6e2d-44b4-8070-ec9084f9f4d4
73009 ERROR  [aux1:a1e5c640-67a0-11ea-93a2-b332b9300b0a:a1f35aec-67a0-11ea-93a2-b332b9300b0a] i.c.s.RepairRunner - Executing SegmentRunner failed
73010 java.lang.AssertionError: Could not release lead on segment a1f35aec-67a0-11ea-93a2-b332b9300b0a
73011     at io.cassandrareaper.storage.CassandraStorage.releaseLead(CassandraStorage.java:1440)
73012     at io.cassandrareaper.service.SegmentRunner.releaseLead(SegmentRunner.java:1226)
73013     at io.cassandrareaper.service.SegmentRunner.run(SegmentRunner.java:183)
73014     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
73015     at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:117)
73016     at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:38)
73017     at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:77)
73018     at com.codahale.metrics.InstrumentedScheduledExecutorService$InstrumentedRunnable.run(InstrumentedScheduledExecutorService.java:241)
73019     at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
73020     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
73021     at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
73022     at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
73023     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
73024     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
73025     at java.lang.Thread.run(Thread.java:748)
73026 INFO   [aux1:a1e5c640-67a0-11ea-93a2-b332b9300b0a:a1edb58c-67a0-11ea-93a2-b332b9300b0a] i.c.s.RepairRunner - postponed repair segment a1edb58c-67a0-11ea-93a2-b332b93      00b0a because segment a1f35aec-67a0-11ea-93a2-b332b9300b0a is running on host 54.148.20.206
```

```
[prod-us-east-1]cpowell@cassandra-reaper-n01:~$ spreaper list-segments -q a1e5c640-67a0-11ea-93a2-b332b9300b0a | jq '.[] | select( .id == "a1f35aec-67a0-11ea-93a2-b332b9300b0a")'
{
  "coordinatorHost": "54.148.20.206",
  "endTime": null,
  "failCount": 0,
  "id": "a1f35aec-67a0-11ea-93a2-b332b9300b0a",
  "repairUnitId": "a11a11d0-67a0-11ea-93a2-b332b9300b0a",
  "runId": "a1e5c640-67a0-11ea-93a2-b332b9300b0a",
  "startTime": 1584375226360,
  "state": "STARTED",
  "tokenRange": {
    "baseRange": {
      "end": -3610345273252933600,
      "start": -3611867519268160500
    },
    "tokenRanges": [
      {
        "end": -3610345273252933600,
        "start": -3611867519268160500
      }
    ]
  }
}
```

You can see the `startTime` of the segment is only a few minutes behind the `start_time` of the repair.
```
[prod-us-east-1]cpowell@cassandra-reaper-n01:~$ spreaper list-runs
# Listing repair runs
# Found 1 repair runs
[
  {
    "blacklisted_tables": [],
    "cause": "cpowell",
    "cluster_name": "aux1",
    "column_families": [
    <sanitzied output>
    ],
    "creation_time": "2020-03-16T16:10:19Z",
    "current_time": "2020-03-16T16:51:35Z",
    "datacenters": [],
    "duration": "40 minutes 15 seconds",
    "end_time": null,
    "estimated_time_of_arrival": "2020-03-17T21:58:07Z",
    "id": "a1e5c640-67a0-11ea-93a2-b332b9300b0a",
    "incremental_repair": false,
    "intensity": 0.900,
    "keyspace_name": "<sanitized_output>",
    "last_event": "postponed repair segment a1efd87d-67a0-11ea-93a2-b332b9300b0a because segment a1f35aec-67a0-11ea-93a2-b332b9300b0a is running on host 54.148.20.206",
    "nodes": [],
    "owner": "cpowell",
    "pause_time": null,
    "repair_parallelism": "PARALLEL",
    "repair_thread_count": 4,
    "repair_unit_id": "a11a11d0-67a0-11ea-93a2-b332b9300b0a",
    "segments_repaired": 145,
    "start_time": "2020-03-16T16:11:20Z",
    "state": "RUNNING",
    "total_segments": 6435
  }
]
```

**Comment by powellchristoph** [comment]
The keyspace being repaired is only replicated to 2 of the 4 datacenters configured in the cluster.

Nodetool status for the keyspace being repaired.
https://gist.github.com/powellchristoph/9b7c87685bc4e9815884d43af040d85e

```
[prod-us-east-1]cpowell@cassandra-aux-n01:~$ nodetool describecluster
Cluster Information:
	Name: Aux 1
	Snitch: org.apache.cassandra.locator.GossipingPropertyFileSnitch
	DynamicEndPointSnitch: disabled
	Partitioner: org.apache.cassandra.dht.Murmur3Partitioner
	Schema versions:
		08be3131-7226-309b-9d2c-8d6e99dd535a: [34.238.7.108, 44.231.2.229, 3.94.151.221, 52.11.136.143, 52.45.134.41, 52.39.219.34, 52.206.199.198, 35.153.147.147, 44.232.84.98, 34.202.217.65, 34.234.34.228, 44.229.253.107, 35.153.246.9, 34.200.45.89, 34.236.55.253, 52.88.9.74, 34.234.48.121, 52.89.250.207, 35.168.37.181, 34.214.81.200, 52.34.73.131, 34.239.41.204, 52.20.102.55, 52.42.10.14, 52.43.196.2, 54.236.89.71, 35.162.166.30, 52.37.84.152, 44.232.132.215, 3.226.27.35, 52.88.126.153, 34.233.186.43, 34.236.92.46, 52.6.92.68, 44.229.246.160, 52.37.100.96, 34.198.73.0, 34.228.246.34, 52.43.211.237, 100.24.84.223, 34.216.185.148, 34.235.52.165, 34.216.181.151, 34.236.16.163, 44.232.209.167, 100.20.82.197, 52.4.51.209, 34.238.91.180, 54.148.20.206, 52.204.208.145, 34.232.27.8, 34.194.25.166, 44.228.202.129, 34.211.71.189, 52.36.23.74, 3.224.247.147, 34.216.39.33, 52.36.7.221, 44.231.240.29, 44.232.192.21]
```

**Comment by powellchristoph** [comment]
If I pause the repair, the segment stays in the `STARTED` state and attempts to abort it also fail.

**Comment by Gappa88** [comment]
@therb1 can you help us?

**Comment by therb1** [comment]
Unfortunately, what you encountered is similar to what I had, but not this.

**Comment by hrast01** [comment]
I'm having a similar/same issue with 2.0.3, after having 1.4.4 running against the same cluster(s) for quite awhile.  I'm also seeing huge CPU utilization as well.

**Comment by adejanovski** [comment]
I'll be looking into this shortly.
Seems indeed like some segments could get stuck in `STARTED` state, subsequently blocking repairs. 
I think this PR should fix it if indeed it's what's happening: https://github.com/thelastpickle/cassandra-reaper/pull/851

**Comment by adejanovski** [comment]
Fixed in the upcoming v2.0.4 (currently being released) through #884
